src/format/parse.rs

```diff
@@ -224,6 +225,263 @@ fn parse_rfc3339<'a>(parsed: &mut Parsed, mut s: &'a str) -> ParseResult<(&'a st
+
+    macro_rules! try_consume {
+        ($e:expr) => {{
+            let (s_, v) = $e?;
+            s = s_;
+            v
+        }};
+    }
+
+    // We assume the maximum number of digits to be 14 and ignore any digits after that.
+    // 10^14 * 3600 fits nicely within an `i64` without overflowing.
+    const MAX_FRACTION_DIGITS: usize = 14;
+    let mut format = Unknown;
+
+    let hour = try_consume!(scan::number(s, 2, 2));
+    parsed.set_hour(if hour == 24 { 0 } else { hour })?;
+
+    if let Some(&b',' | &b'.') = s.as_bytes().first() {
+        // Minute, second and nanosecond are expressed as a fraction of an hour.
+        s = &s[1..];
+        let digits = s.as_bytes().iter().take_while(|c| c.is_ascii_digit()).count();
+        let (_, fraction) = scan::number(s, 1, MAX_FRACTION_DIGITS)?;
+        s = &s[digits..];
+        let (seconds, nanos) =
+            fraction_to_seconds(fraction, min(digits, MAX_FRACTION_DIGITS), 3600);
+
+        parsed.set_minute(seconds / 60)?;
+        parsed.set_second(seconds % 60)?;
+        parsed.set_nanosecond(nanos)?;
+        return Ok((s, format, hour == 24));
+    }
+
+    let c = s.as_bytes().first().unwrap_or(&b'a');
+    if !(c.is_ascii_digit() || c == &b':') {
+        // Allow reduced accuracy
+        parsed.set_minute(0)?;
+        return Ok((s, format, hour == 24));
+    }
+
+    format = if s.as_bytes().first() == Some(&b':') { Extended } else { Basic };
+    if format == Extended {
+        s = &s[1..];
+    }
+    parsed.set_minute(try_consume!(scan::number(s, 2, 2)))?;
+
+    if let Some(&b',' | &b'.') = s.as_bytes().first() {
+        // Second and nanosecond are expressed as a fraction of a minute.
+        s = &s[1..];
+        let digits = s.as_bytes().iter().take_while(|c| c.is_ascii_digit()).count();
+        let (_, fraction) = scan::number(s, 1, MAX_FRACTION_DIGITS)?;
+        s = &s[digits..];
+        let (seconds, nanos) = fraction_to_seconds(fraction, min(digits, MAX_FRACTION_DIGITS), 60);
+
+        parsed.set_second(seconds)?;
+        parsed.set_nanosecond(nanos)?;
+        return Ok((s, format, hour == 24));
+    }
+
+    let c = s.as_bytes().first().unwrap_or(&b'a');
+    if !(c.is_ascii_digit() || (format == Extended && c == &b':')) {
+        // Allow reduced accuracy
+        return Ok((s, format, hour == 24));
+    }
+
+    if format == Extended {
+        s = scan::char(s, b':')?;
+    }
+    parsed.set_second(try_consume!(scan::number(s, 2, 2)))?;
+
+    if let Some(&b',' | &b'.') = s.as_bytes().first() {
+        let nano = try_consume!(scan::nanosecond(&s[1..]));
+        parsed.set_nanosecond(nano)?;
+    }
+
+    if hour == 24
+        && (parsed.minute.unwrap_or(0) != 0
+            || parsed.second.unwrap_or(0) != 0
+            || parsed.nanosecond.unwrap_or(0) != 0)
+    {
+        return Err(INVALID);
+    }
+    Ok((s, format, hour == 24))
+}
+
+fn fraction_to_seconds(fraction: i64, digits: usize, max_seconds: i64) -> (i64, i64) {
```

Comment:
Please docstring comment the meaning of the returned tuple value.

    /// Returns the passed values as seconds and nanoseconds.